helm/data_repository.go

```diff
@@ -120,6 +120,8 @@ func getRepository(d *schema.ResourceData, m *Meta) (*repo.Entry, error) {
 }
 
 func addRepository(m *Meta, name, url string, home helmpath.Home, certFile, keyFile, caFile string, username string, password string) error {
+	m.Lock()
```

Comment:
Locking can probably happen later when reading from and writing to the repositories file is actually required.
This would allow downloading indexes in parallel.

The result of the `LoadRepositoriesFile()` call stored in `repoFile` is only needed after the index caching for `repoFile.Update()` at https://github.com/terraform-providers/terraform-provider-helm/pull/272/files#diff-a64ebd9d5beb4bea06b92cff0bddf4c7R152

The original helm implementation also checks if the repo does not already exists and doesn't need an update: https://github.com/helm/helm/blob/518a9d63e6cdcc06a816c6b6ca27ff36fe1636ec/cmd/helm/repo_add.go#L109
It may be useful to add an equivalent check before locking and rewriting the repositories file in order to skip no-op updates and reduce contention.

Note: locking only at the process level won't prevent concurrent processes from overwriting each others repositories additions when they share the same helm home. Process level mutex locking could be combined with file level locking to handle this.

Reply:
Using github.com/gofrs/flock for cross platform file locking for same process and concurrent processes eases that task: https://github.com/helm/helm/pull/5678/files#diff-13a230f2627b662a7d5319ba6c0d653cR139

Reply:
I do agree that current implementation the issue of downloading the indexes every time and this fix won't make things faster.

Since we have a bug, what do you think about merging #273 here, merging this PR into master and release a new version with the fixes? I would create an issue with your suggestions for a better implementation so that it's worked while we have a less buggy version out there

Reply:
Agreed, let's do this.

Reply:
My PR implementing the fix using github.com/gofrs/flock on helm itself was just merged: https://github.com/helm/helm/pull/5678

Reply:
Feel free to merge https://github.com/terraform-providers/terraform-provider-helm/pull/273 here.